controllers/delete.go

```diff
@@ -20,18 +20,29 @@ import (
 	"context"
 	"fmt"
 
+	ibmv1alpha1 "github.com/IBM/ibm-storage-odf-operator/api/v1alpha1"
 	"github.com/go-logr/logr"
+	ocsv1 "github.com/openshift/ocs-operator/api/v1"
+	odfv1alpha1 "github.com/red-hat-data-services/odf-operator/api/v1alpha1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/types"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-
-	ibmv1alpha1 "github.com/IBM/ibm-storage-odf-operator/api/v1alpha1"
-	ocsv1 "github.com/openshift/ocs-operator/api/v1"
-	odfv1alpha1 "github.com/red-hat-data-services/odf-operator/api/v1alpha1"
 )
 
 func (r *StorageSystemReconciler) deleteResources(instance *odfv1alpha1.StorageSystem, logger logr.Logger) error {
 
+	var storageSystems odfv1alpha1.StorageSystemList
+	err := r.Client.List(context.TODO(), &storageSystems)
+	if err != nil {
+		// Log the error but not fail the operator
+		logger.Error(err, "Failed to List", "Kind", instance.Spec.Kind)
+	}
+	// Check for number of StorageSystems in the cluster, If it is 2 or more that means StorageSystems will remain after deletion of current instance. If it is less than or equal to 1 then proceed with deleting all quickstarts
+	if err == nil && len(storageSystems.Items) <= 1 {
+		// Attempt to delete quickstarts if present. Otherwise log and move on
+		r.deleteQuickStarts(logger)
+	}
+
```

Comment:
This logic should not be here. Create a separate generic func for this in the same file which can be used at other places as well. As you are checking the list can you also check if all of them are marked for deletion?

pls add tests also

Reply:
I've created a separate function for the storagesystem and added tests